### ISSUE TITLE
feat(brain): the extraction cascade — protect the review gate before M3 source breadth

### ISSUE THREAD

**Issue by msywulak** [AtlasDevHQ/atlas#5334]
ADR-0036 §Consequences already names this as an open obligation:

> The human review gate is now load-bearing beyond the KB: it is the conflict-resolution mechanism (T4) and the write-back exit (T9). **Its throughput and UX become a first-class concern the moment M1 ships facts at connector scale.**

M1 ships one chat connector. **M3 is source breadth** — transcripts → email → docs/wiki/code/drive. That is the moment the episode firehose multiplies, and today every episode drained by `runBrainExtractionCycle` costs one `generateObject` call against a frontier model regardless of whether it contains a claim. This issue is that concern made concrete, plus the architecture that answers it.

## Where the models actually sit — read this first

The recurring confusion is that the fine-tuned model *is* the brain. It is not. **The facts never become a model; they stay rows.**

```
episodes → [Model A: small, local] → candidate facts → HUMAN GATE → published facts (Postgres)
                                                                          ↓
                                                              searchBrain retrieves them
                                                                          ↓
user question + retrieved facts + semantic layer → [Model B: frontier] → answer
```

| | **Model A** (ingest) | **Model B** (interactive) |
|---|---|---|
| Job | episode text → proposed SPO | question + facts → answer |
| Runs | CPU, batch, hours | frontier, sub-second |
| Sees a user? | never | always |
| Trained? | yes — on public + first-party data (#5339) | no |
| Knows any company facts? | **no** | no — they arrive in its context |

Model A is a data-entry clerk. It reads one message and proposes a claim. It never answers a question, never sees the fact graph, never talks to a person. **This whole arc is about Model A.**

Atlas gets smarter because **Model B's context gets better** — `searchBrain` hands it grounded, attributed, grant-scoped facts at query time. Not because any model learned them. That is why rows rather than weights: revocation, provenance, per-fact grants, as-of reads and surviving disagreement are all properties of a table and none of them are properties of a checkpoint (#5339).

## The reframe that orders this arc

The instinct is to treat volume as an inference-cost problem and solve it with a cheaper model. That is the second-order problem. The first-order one:

- ~100k chat messages/month for a mid-size company; Teams + email + transcripts multiply that 3–5×.
- Even at a **2% candidate rate**, that is a few thousand drafts/month — **~100/working-day** for a human to review.
- There is no setting that turns the gate off. PRD commitment 3 ("will not learn behind your back") and finish condition 2 ("every authoritative claim has a human name on it") both rest on it.

So the scarce resource is human review attention, and it cannot be bought. That changes the optimization target: **precision at the gate, not recall from the extractor**.

## The boundary: interactive stays frontier, ingest gets optimised

**Model tier follows the latency budget and the blast radius, not the subsystem.**

| | Interactive (Model B) | Ingest (Model A) |
|---|---|---|
| Latency budget | sub-second first token | hours |
| Volume | ~10k calls/month | ~100k+/month |
| A wrong output reaches | a person, as an answer | a review queue, as a draft |
| Batch API | impossible | ideal (#5352) |
| Local model | never | eventually (#5337) |

Every slice in this arc is **ingest-only**. `agent.ts` already recovers most of the interactive cost through `cacheControl: ephemeral` (~80% on steps 2+). Precedent for tiering by job rather than subsystem already exists in `settings.ts`, which carries a cheaper compaction-model override.

## The cascade

| Stage | Job | Model | Runs on |
|---|---|---|---|
| 0 | Deterministic routing — length, channel scope, thread position, known-negative shapes | none | free |
| 1 | "Does this episode contain a promotable claim?" | fine-tuned encoder, 20–150M (MiniLM / ModernBERT class), int8 | **CPU, trivially** |
| 2 | Episode → `FactCandidate`, constrained decode against `ExtractionSchema` | distilled 1–4B, Q4/Q8 | **CPU, comfortably** |
| 3 | Ambiguous episodes; tension judgment | frontier | API |

Stage 1 is the whole game, and it is the part CPU is *best* at: an encoder classifier is prefill-only, no autoregressive decode. Int8 MiniLM-class on 8 cores handles high thousands of short episodes/sec. 90–95% of the firehose dies there.

Stage 2 latency is fine because **extraction is already a batch job by construction**: `extract.ts` is a separate fiber precisely so "episode freshness never blocks on LLM latency / 429s", and the backlog is visible in `idx_brain_episodes_extraction_queue`.

## The economics, measured

100-employee workspace, ~100k episodes/month, ~520 input / 60 output tokens per extraction call. Three-year Slack backfill ≈ 2M messages.

| | Backfill | Steady state |
|---|---|---|
| Opus-tier, synchronous, unfiltered | ~$7,000 | ~$410/mo |
| + Haiku tier (#5353) | ~$1,400 | ~$82/mo |
| + Batch API (#5352) | ~$700 | ~$41/mo |
| + triage (#5336) | ~$70 | ~$4–8/mo |

**Two days of work captures ~99% of the saving.** #5352 and #5353 come first for that reason, and #5337 is explicitly *not* justified on token price — see its trigger conditions.

**Prompt caching does not apply here.** The fixed prefix (`EXTRACTION_SYSTEM_PROMPT` ~150 tokens + serialized `ExtractionSchema` ~200) is well under the ~1024-token minimum cacheable prefix. It would silently never cache.

## Two seams already exist

- **`FactExtractor` (`extract.ts:438`)** — *"a later, better extractor replaces this one function without the cycle knowing."*
- **`providers.ts` already supports `ollama` and `openai-compatible`**, both keyless by design.

## What Model A may never do

- **Grants are never generated.** `visible_to` is derived deterministically and widened at publish by `widenGrantFromEvidence`. Commitment 7: over-restriction is the acceptable failure; generation has no such bias.
- **Provenance is a pointer, not a generation.**
- **Entity resolution stays structural.** `reconcile.ts` and `entity-store.ts` prohibit read-time fuzzy/embedding/LLM disambiguation; `alias-proposal.ts` prohibits resemblance as a proposal source, falsified on this repo's own corpus — `led_by` and `leads` are inverse relations and the top-ranked pair any similarity detector returns.
- **Tension detection escalates to stage 3.** Commitment 2 says recency does not get to win.

## Why this earns a slot

Per the PRD, a milestone advancing none of the eight finish conditions is worth questioning. This advances two:

- **Condition 8 (the self-hoster has all of it).** A self-host that needs a frontier key — or a GPU — to ingest Slack fails this in practice even if it passes on paper.
- **Condition 6 (the limits are visible and honest).** Stage 1 introduces a new way to be silently wrong: a dropped real claim. ADR-0041's coverage surface has to be honest about what Atlas can see.

## Route

1. #5339 — the ADR (two prohibitions). Zero dependencies, writable today.
2. #5335 — gate-decision export, scoped to evaluation and analytics.
3. #5352 — Batch API. ~50% off a fiber already built for it.
4. #5353 — extraction model tier. ~5× on a call nobody sees unattended.
5. #5336 stage 0 — deterministic routing, needs no model.
6. #5338 — freeze the held-out set *before* anything is trained.
7. #5336 stage 1, then #5337.

Adjacent and unblocked: #5354 — strip quoted reply chains before extraction. Cheap; its payoff arrives with M3 email volume.

Full rationale and open questions: the wayfinder map at #5343.

## Key files

- `packages/api/src/lib/brain/extract.ts` — `FactExtractor` seam (438), `llmFactExtractor` (446), `resolveExtractionModel` (586), `runBrainExtractionCycle` (804), `BATCH_SIZE`, quarantine ledger
- `packages/api/src/lib/brain/candidates.ts` — `EPISODE_BODY_MAX_CHARS`, candidate reads
- `packages/api/src/lib/brain/promotion.ts` — the gate; `classifyFactForPromotion`
- `packages/api/src/lib/brain/search.ts`, `fusion.ts` — how published facts reach Model B
- `packages/api/src/lib/brain/reconcile.ts`, `entity-store.ts`, `alias-proposal.ts` — the prohibitions this arc must not violate
- `packages/api/src/lib/brain/ingest/outlook/client.ts` — `composeEmailBody`, the #5354 surface
- `packages/api/src/lib/providers.ts` — `ollama` / `openai-compatible` already present
- `packages/api/src/lib/agent.ts` — Model B; this arc must not touch it
- `packages/api/src/lib/db/migrations/0180_brain_substrate.sql` — `source_episode_id` composite FK, the extraction-queue index
- `packages/api/src/lib/learn/pattern-tiers.ts`, `promote-decay.ts` — in-tree precedent for cheap-tier-gates-expensive-tier
- `packages/api/src/lib/scheduler/` — `registerPeriodicFiber`, where stage 0/1 run
- `docs/prd/company-atlas.md`, `docs/adr/0036-atlas-as-company-brain.md`, `docs/adr/0041-the-coverage-surface-counts-what-it-can-see.md`, `docs/agents/practices.md`

## Acceptance criteria

- [ ] The arc lands as independently shippable slices (see Dependencies), and the cascade degrades to today's behaviour when every stage but 2 is disabled.
- [ ] #5352 and #5353 ship first — two days, ~99% of the cost saving.
- [ ] Stage 0 + stage 1 ship **before** M3 opens a second source class, so breadth never hits an unfiltered gate.
- [ ] Model B (the interactive path) is provably untouched by every slice in this arc.
- [ ] A self-hosted install with `ATLAS_PROVIDER=ollama` and no frontier key extracts facts end-to-end on CPU, and PRD conditions 1–7 still hold on it (condition 8).
- [ ] No stage generates a grant, a provenance pointer, or an entity equivalence.
- [ ] The measurement slice reports stage-1 recall and stage-2 gate agreement, and both can come back bad.

## Dependencies

- #5339 — ADR: the two training prohibitions
- #5335 — gate-decision export (evaluation corpus)
- #5352 — Batch API for the extraction fiber
- #5353 — extraction model tier, ingest-only
- #5336 — stage 0/1 pre-extraction triage
- #5338 — the failing-capable measurement
- #5337 — stage 2 distilled CPU extractor
- #5354 — quoted-reply stripping (adjacent, unblocked)
- #5343 — wayfinder map

Blocked by nothing. Blocks M3 source breadth in practice, though not in the tree.

**Comment by msywulak** [comment]
Closing as part of the hosted-SaaS shutdown (ADR-0047, #5679). The backlog is being reset; any future work will be re-planned from scratch around Atlas as a purely self-deployable project. Reopen or re-file if this still applies there.
